qiskit/transpiler/layout.py

```diff
@@ -738,3 +743,218 @@ def final_virtual_layout(self, filter_ancillas: bool = True) -> Layout:
+        # an initial layout, even if there isn't actually any laying out to hardware.
+        if initial_layout is None:
+            initial_layout = Layout(dict(enumerate(dag.qubits)))
+        if final_layout is None:
+            final_layout = Layout(dict(enumerate(dag.qubits)))
+
+        input_qubits = sorted(input_qubit_indices, key=input_qubit_indices.get)
+
+        num_qubits = len(dag.qubits)
+
+        # Throughout the rest of this, we will speak about index permutations as lists that mean:
+        #
+        #    qubit `permutation[i]` goes to new index `i`
+        #
+        # or in alternative langauge,
+        #
+        #   after the permutation, qubit `i` contains qubit `permutation[i]`.
+        #
+        # This is to match the convention that `PermutationGate` uses, but beware: it might not be
+        # the way you think about permutations (it's not my preferred convention---Jake).
+        #
+        # Now, we'll step through the tranpsilation process.  At each point, we'll relate the
+        # objects we have back to a 3-tuple of abstract objects, which are applied in order:
+        #
+        #   (relabelling, explicit instructions, implicit instructions)
+        #
+        # The "explicit instructions" are always just the DAG itself.  The "relabelling" is
+        # generally associated with the "initial layout" and the metadata linking the original
+        # virtual qubit objects and their indices.  The "implicit instructions" is where all the
+        # interesting stuff happens; at the moment, in Qiskit, we only track an implicit final
+        # permutation, though you could imagine a world where we allow a lot more things to be
+        # tracked, such as necessary classical post-processing steps.
+        #
+        # We will attempt to always have in hand an "undoing" permutation, such that doing
+        # `qc.append(PermutationGate(permutation), qc.qubits)` to the output of the transpiler, were
+        # it to terminate at any given step, would be enough to precisely recreate the unitary of
+        # the circuit (with due hand-waviness around measurements/resets), up to the qubit
+        # relabelling of the initial layout.
+
+        # First, virtual permutation modifications happen.  For example, `ElidePermutations` or
+        # `StarPreRouting`.  Note that `virtual_permutation_layout` uses an opposite convention to
+        # `final_layout` for defining the permutation.
+        undo_elided_on_virtuals = [
+            virtual_permutation_layout[virtual_bit]
+            for virtual_bit in input_qubits[:num_input_qubits]
+        ]
+        # `virtual_permutation_layout` is defined without ancillas.  If they got added later, extend
+        # the virtual permutation with the implicit identity on the other components.
+        if num_qubits > len(undo_elided_on_virtuals):
+            undo_elided_on_virtuals.extend(range(len(undo_elided_on_virtuals), num_qubits))
```

Comment:
Did this happen in your testing? I'd be surprised if so, because we should expand with ancillas after anything that introduces a virtual permutation and I would expect the ancilla expansion or layout application to transform the property set.

Reply:
> I would expect the ancilla expansion or layout application to transform the property set.

That doesn't happen.  Amazingly, I think in the parent of this PR, it's still `ApplyLayout` that "expands" `original_qubit_indices` instead of, you know, the pass that did the expansion.

Reply:
Ah sorry it was a typo, I meant to say I **wouldn't** expect. That matches my expectation and memory of how those passes work.